Metabase will not let you update the cert if it was rotated

### Describe the bug

Snowflake will return an error if updating certs when these were rotated on the source system

### To Reproduce

1) create a user on snowflake with certs, just certs, no password. Follow https://docs.snowflake.com/en/user-guide/key-pair-auth
2) add the db to Metabase, remember that it's a user just with certs
3) then go to the user in Snowflake and rotate the key (add a new Public key)
4) update the private key in Metabase

### Expected behavior

We should disconnect the DB when doing the conn update

### Logs

```
2023-08-23 23:22:02,255 ERROR driver.util :: Failed to connect to Database
net.snowflake.client.jdbc.SnowflakeSQLException: JWT token is invalid.
	at net.snowflake.client.core.SessionUtil.newSession(SessionUtil.java:695)
	at net.snowflake.client.core.SessionUtil.openSession(SessionUtil.java:291)
	at net.snowflake.client.core.SFSession.open(SFSession.java:477)
	at net.snowflake.client.jdbc.DefaultSFConnectionHandler.initialize(DefaultSFConnectionHandler.java:104)
	at net.snowflake.client.jdbc.DefaultSFConnectionHandler.initializeConnection(DefaultSFConnectionHandler.java:79)
	at net.snowflake.client.jdbc.SnowflakeConnectionV1.initConnectionWithImpl(SnowflakeConnectionV1.java:116)
	at net.snowflake.client.jdbc.SnowflakeConnectionV1.<init>(SnowflakeConnectionV1.java:96)
	at net.snowflake.client.jdbc.SnowflakeDriver.connect(SnowflakeDriver.java:180)
	at metabase.plugins.jdbc_proxy$proxy_driver$reify__79655.connect(jdbc_proxy.clj:37)
	at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
	at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:382)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:302)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:299)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__50188.invoke(connection.clj:311)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__50170.invoke(connection.clj:287)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:285)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:281)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:310)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:306)
	at metabase.driver.sql_jdbc$fn__111540.invokeStatic(sql_jdbc.clj:43)
	at metabase.driver.sql_jdbc$fn__111540.invoke(sql_jdbc.clj:41)
	at metabase.driver.snowflake$fn__107979.invokeStatic(snowflake.clj:508)
	at metabase.driver.snowflake$fn__107979.invoke(snowflake.clj:506)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__46898.invoke(util.clj:144)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2023-08-23 23:22:02,256 ERROR driver.snowflake :: (type message)
=> java.lang.String
2023-08-23 23:22:02,257 ERROR api.database :: Cannot connect to Database
clojure.lang.ExceptionInfo: JWT token is invalid. {:message "JWT token is invalid."}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:160)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:132)
	at clojure.lang.RestFn.invoke(RestFn.java:442)
	at metabase.api.database$test_database_connection.invokeStatic(database.clj:678)
	at metabase.api.database$test_database_connection.doInvoke(database.clj:668)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at metabase.api.database$fn__96187.invokeStatic(database.clj:888)
	at metabase.api.database$fn__96187.invoke(database.clj:864)
	at compojure.core$wrap_response$fn__42432.invoke(core.clj:160)
	at compojure.core$wrap_route_middleware$fn__42416.invoke(core.clj:132)
	at compojure.core$wrap_route_info$fn__42421.invoke(core.clj:139)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:151)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:153)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:153)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at metabase.server.middleware.auth$enforce_authentication$fn__87846.invoke(auth.clj:17)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at compojure.core$make_context$handler__42472.invoke(core.clj:290)
	at compojure.core$make_context$fn__42476.invoke(core.clj:300)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:199)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:199)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:199)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at metabase.api.routes$fn__99658$fn__99659.invoke(routes.clj:62)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.server.routes$fn__99819$fn__99820.doInvoke(routes.clj:70)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at compojure.core$make_context$handler__42472.invoke(core.clj:290)
	at compojure.core$make_context$fn__42476.invoke(core.clj:300)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:153)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:153)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__42425.invoke(core.clj:153)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:199)
	at compojure.core$routes$fn__42444$f__42445$respond_SINGLEQUOTE___42446.invoke(core.clj:197)
	at compojure.core$make_context$fn__42476.invoke(core.clj:301)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at compojure.core$routes$fn__42444$f__42445.invoke(core.clj:198)
	at compojure.core$routes$fn__42444.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__96459.invoke(exceptions.clj:103)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__96456.invoke(exceptions.clj:91)
	at metabase.server.middleware.log$log_api_call$fn__103115$fn__103116$fn__103117.invoke(log.clj:216)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:18)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__103115$fn__103116.invoke(log.clj:208)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__103115.invoke(log.clj:207)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__109567.invoke(browser_cookie.clj:42)
	at metabase.server.middleware.security$add_security_headers$fn__80396.invoke(security.clj:153)
	at metabase.server.middleware.json$wrap_json_body$fn__106819.invoke(json.clj:67)
	at metabase.server.middleware.offset_paging$handle_paging$fn__80420.invoke(offset_paging.clj:45)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__106837.invoke(json.clj:103)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__109834.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__109853.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__50568.invoke(misc.clj:61)
	at metabase.server.middleware.session$reset_session_timeout$fn__63848.invoke(session.clj:441)
	at metabase.server.middleware.session$bind_current_user$fn__63817$fn__63818.invoke(session.clj:336)
	at metabase.server.middleware.session$do_with_current_user.invokeStatic(session.clj:315)
	at metabase.server.middleware.session$do_with_current_user.invoke(session.clj:304)
	at metabase.server.middleware.session$bind_current_user$fn__63817.invoke(session.clj:335)
	at metabase.server.middleware.session$wrap_current_user_info$fn__63798.invoke(session.clj:285)
	at metabase.server.middleware.session$wrap_session_id$fn__63781.invoke(session.clj:217)
	at metabase.server.middleware.auth$wrap_api_key$fn__87854.invoke(auth.clj:30)
	at ring.middleware.cookies$wrap_cookies$fn__109754.invoke(cookies.clj:216)
	at metabase.server.middleware.misc$add_content_type$fn__50550.invoke(misc.clj:29)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__50576.invoke(misc.clj:78)
	at ring.middleware.gzip$wrap_gzip$fn__109796.invoke(gzip.clj:86)
	at metabase.server.middleware.misc$bind_request$fn__50579.invoke(misc.clj:95)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__109583.invoke(ssl.clj:41)
	at metabase.server$async_proxy_handler$fn__101881.invoke(server.clj:77)
	at metabase.server.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:416)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:385)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:272)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:140)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:936)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1080)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: net.snowflake.client.jdbc.SnowflakeSQLException: JWT token is invalid.
	at net.snowflake.client.core.SessionUtil.newSession(SessionUtil.java:695)
	at net.snowflake.client.core.SessionUtil.openSession(SessionUtil.java:291)
	at net.snowflake.client.core.SFSession.open(SFSession.java:477)
	at net.snowflake.client.jdbc.DefaultSFConnectionHandler.initialize(DefaultSFConnectionHandler.java:104)
	at net.snowflake.client.jdbc.DefaultSFConnectionHandler.initializeConnection(DefaultSFConnectionHandler.java:79)
	at net.snowflake.client.jdbc.SnowflakeConnectionV1.initConnectionWithImpl(SnowflakeConnectionV1.java:116)
	at net.snowflake.client.jdbc.SnowflakeConnectionV1.<init>(SnowflakeConnectionV1.java:96)
	at net.snowflake.client.jdbc.SnowflakeDriver.connect(SnowflakeDriver.java:180)
	at metabase.plugins.jdbc_proxy$proxy_driver$reify__79655.connect(jdbc_proxy.clj:37)
	at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
	at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:382)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:302)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:299)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__50188.invoke(connection.clj:311)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__50170.invoke(connection.clj:287)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:285)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:281)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:310)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:306)
	at metabase.driver.sql_jdbc$fn__111540.invokeStatic(sql_jdbc.clj:43)
	at metabase.driver.sql_jdbc$fn__111540.invoke(sql_jdbc.clj:41)
	at metabase.driver.snowflake$fn__107979.invokeStatic(snowflake.clj:508)
	at metabase.driver.snowflake$fn__107979.invoke(snowflake.clj:506)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__46898.invoke(util.clj:144)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
2023-08-23 23:22:02,259 DEBUG middleware.log :: PUT /api/database/4 400 146.1 ms (4 DB calls) 
{:message "JWT token is invalid."}
```

### Information about your Metabase installation

```JSON
46.7+
```


### Severity

P2

### Additional context

_No response_

--- a/src/metabase/api/database.clj
+++ b/src/metabase/api/database.clj
@@ -880,7 +880,7 @@
                                    (upsert-sensitive-fields existing-database))
         ;; verify that we can connect to the database if `:details` OR `:engine` have changed.
         details-changed?  (some-> details (not= (:details existing-database)))
-        engine-changed?   (some-> engine (not= (:engine existing-database)))
+        engine-changed?   (some-> engine keyword (not= (:engine existing-database)))
         conn-error        (when (or details-changed? engine-changed?)
                             (test-database-connection (or engine (:engine existing-database))
                                                       (or details (:details existing-database))))

--- a/src/metabase/models/secret.clj
+++ b/src/metabase/models/secret.clj
@@ -203,11 +203,21 @@
   [details secret-property]
   (let [{path-kw :path, value-kw :value, options-kw :options, id-kw :id} (get-sub-props secret-property)
         id (id-kw details)
-        value (if id
-                (String. ^bytes (:value (t2/select-one Secret :id id)) "UTF-8")
-                (value-kw details))]
+        ;; When a secret is updated, we get both a new value as well as the ID of old secret.
+        value (or (when-let [value (value-kw details)]
+                    (if (string? value)
+                      value
+                      (String. ^bytes value "UTF-8")))
+                  (when id
+                    (String. ^bytes (:value (t2/select-one Secret :id id)) "UTF-8")))]
     (case (options-kw details)
-      "uploaded" (String. ^bytes (driver.u/decode-uploaded value) "UTF-8")
+      "uploaded" (try
+                   ;; When a secret is updated, the value has already been decoded
+                   ;; instead of checking if the string is base64 encoded, we just
+                   ;; try to decoded it and leave it as is if the attempt fails.
+                   (String. ^bytes (driver.u/decode-uploaded value) "UTF-8")
+                   (catch IllegalArgumentException _
+                     value))
       "local" (slurp (if id value (path-kw details)))
       value)))
 

--- a/test/metabase/models/secret_test.clj
+++ b/test/metabase/models/secret_test.clj
@@ -75,27 +75,51 @@
              (secret/get-secret-string
               {:secret-prop-id      id
                :secret-prop-options "uploaded"}
-              "secret-prop")))))
+              "secret-prop")))
+      (testing "but prefer value if both value and id are given (#33452)"
+        (are [value] (= "psszt!"
+                        (secret/get-secret-string
+                         {:secret-prop-id      id
+                          :secret-prop-value   value
+                          :secret-prop-options "uploaded"}
+                         "secret-prop"))
+          "psszt!"
+          (.getBytes "psszt!" "UTF-8")
+          (let [encoder (java.util.Base64/getEncoder)]
+            (str "data:application/octet-stream;base64,"
+                 (.encodeToString encoder
+                                  (.getBytes "psszt!" "UTF-8"))))))))
 
-  (mt/with-temp-file [file "-key.pem"]
-    (spit file "titok")
-    (testing "get-secret-string from local file in value"
-      (is (= "titok"
-             (secret/get-secret-string
-              {:secret-prop-path    file
-               :secret-prop-options "local"}
-              "secret-prop"))))
+  (testing "get-secret-string from local file"
+    (mt/with-temp-file [file-db "-1-key.pem"
+                        file-value "-2-key.pem"]
+      (spit file-db "titok")
+      (spit file-value "psszt!")
 
-    (testing "get-secret-string from local file in the database"
-      (t2.with-temp/with-temp [Secret {id :id} {:name       "private-key"
-                                                :kind       ::secret/pem-cert
-                                                :value      file
-                                                :creator_id (mt/user->id :crowberto)}]
+      (testing "from value"
         (is (= "titok"
                (secret/get-secret-string
-                {:secret-prop-id      id
+                {:secret-prop-path    file-db
                  :secret-prop-options "local"}
-                "secret-prop")))))))
+                "secret-prop"))))
+
+      (testing "from the database"
+        (t2.with-temp/with-temp [Secret {id :id} {:name       "private-key"
+                                                  :kind       ::secret/pem-cert
+                                                  :value      file-db
+                                                  :creator_id (mt/user->id :crowberto)}]
+          (is (= "titok"
+                 (secret/get-secret-string
+                  {:secret-prop-id      id
+                   :secret-prop-options "local"}
+                  "secret-prop")))
+          (testing "but prefer value if both value and id are given (#33452)"
+            (is (= "psszt!"
+                   (secret/get-secret-string
+                    {:secret-prop-id      id
+                     :secret-prop-value   file-value
+                     :secret-prop-options "local"}
+                    "secret-prop")))))))))
 
 (deftest value->file!-test
   (testing "value->file! works for a secret value"

